osutil/epoll/epoll_test.go

```diff
@@ -0,0 +1,546 @@
+		c.Assert(buf, DeepEquals, msg)
+	}
+
+	err = e.Deregister(listenerFd)
+	c.Assert(err, IsNil)
+
+	err = e.Close()
+	c.Assert(err, IsNil)
+}
+
+func (*epollSuite) TestRegisterMultiple(c *C) {
+	e, err := epoll.Open()
+	c.Assert(err, IsNil)
+
+	numSockets := 20
+
+	socketRxFds := make([]int, 0, numSockets)
+	socketTxFds := make([]int, 0, numSockets)
+
+	msg1 := []byte("foo")
+	msg2 := []byte("bar")
+
+	for i := 0; i < numSockets; i++ {
+		socketFds, err := unix.Socketpair(unix.AF_UNIX, unix.SOCK_STREAM, 0)
+		c.Assert(err, IsNil)
+		defer unix.Close(socketFds[0])
+		defer unix.Close(socketFds[1])
+
+		listenerFd := socketFds[0]
+		senderFd := socketFds[1]
+
+		err = unix.SetNonblock(listenerFd, true)
+		c.Assert(err, IsNil)
+
+		err = e.Register(listenerFd, epoll.Readable)
+		c.Assert(err, IsNil)
+
+		_, err = unix.Write(senderFd, msg1)
+		c.Assert(err, IsNil)
+
+		socketRxFds = append(socketRxFds, listenerFd)
+		socketTxFds = append(socketTxFds, senderFd)
+	}
+
+	for _, senderFd := range socketTxFds {
+		_, err = unix.Write(senderFd, msg2)
+		c.Assert(err, IsNil)
+	}
+
+	events, err := e.Wait()
+	c.Assert(err, IsNil)
+	c.Assert(len(events), Equals, len(socketRxFds))
+
+	for i, listenerFd := range socketRxFds {
+		buf := make([]byte, len(msg1))
+		c.Assert(events[i].Fd, Equals, listenerFd)
+		_, err = unix.Read(events[i].Fd, buf)
+		c.Assert(err, IsNil)
+		c.Assert(buf, DeepEquals, msg1)
+	}
+
+	for i, listenerFd := range socketRxFds {
+		buf := make([]byte, len(msg2))
+		c.Assert(events[i].Fd, Equals, listenerFd)
+		_, err = unix.Read(events[i].Fd, buf)
+		c.Assert(err, IsNil)
+		c.Assert(buf, DeepEquals, msg2)
+	}
+
+	for i := 0; i < len(socketRxFds)/2; i++ {
+		err = e.Deregister(socketRxFds[i])
+		c.Assert(err, IsNil)
+	}
+
+	msg3 := []byte("baz")
+
+	for _, senderFd := range socketTxFds {
+		_, err = unix.Write(senderFd, msg3)
+		c.Assert(err, IsNil)
+	}
+
+	events, err = e.Wait()
+	c.Assert(err, IsNil)
+	c.Assert(len(events), Equals, len(socketRxFds)/2)
+
+	for i, listenerFd := range socketRxFds[len(socketRxFds)/2:] {
+		buf := make([]byte, len(msg3))
+		c.Assert(events[i].Fd, Equals, listenerFd)
+		_, err = unix.Read(events[i].Fd, buf)
+		c.Assert(err, IsNil)
+		c.Assert(buf, DeepEquals, msg3)
+	}
+
+	err = e.Close()
+	c.Assert(err, IsNil)
+}
+
+func (epollSuite) TestRegisterDeregisterConcurrency(c *C) {
+	e, err := epoll.Open()
+	c.Assert(err, IsNil)
+	c.Assert(e.RegisteredFdCount(), Equals, 0)
+
+	concurrencyCount := 20
+
+	listenerFds := make([]int, 0, concurrencyCount)
+
+	for i := 0; i < concurrencyCount; i++ {
+		socketFds, err := unix.Socketpair(unix.AF_UNIX, unix.SOCK_STREAM, 0)
+		c.Assert(err, IsNil)
+		defer unix.Close(socketFds[0])
+		defer unix.Close(socketFds[1])
+
+		listenerFd := socketFds[0]
+
+		err = unix.SetNonblock(listenerFd, true)
+		c.Assert(err, IsNil)
+
+		listenerFds = append(listenerFds, listenerFd)
+	}
+
+	errCh := make(chan error)
+
+	for _, fd := range listenerFds {
+		go concurrentlyRegister(e, fd, errCh)
```

Comment:
(super-nitick) as  concurrentlyRegisteris only used here and is relatively short I wonder if it should just be included here directly:
```go
diff --git a/osutil/epoll/epoll_test.go b/osutil/epoll/epoll_test.go
index 9e0a031b18..7e22d704a8 100644
--- a/osutil/epoll/epoll_test.go
+++ b/osutil/epoll/epoll_test.go
@@ -39,11 +39,6 @@ func (*epollSuite) TestOpenClose(c *C) {
        c.Assert(e.Fd(), Equals, -1)
 }
 
-func concurrentlyRegister(e *epoll.Epoll, fd int, errCh chan error) {
-       err := e.Register(fd, epoll.Readable)
-       errCh <- err
-}
-
 func concurrentlyDeregister(e *epoll.Epoll, fd int, errCh chan error) {
        err := e.Deregister(fd)
        errCh <- err
@@ -407,7 +402,11 @@ func (epollSuite) TestRegisterDeregisterConcurrency(c *C) {
        errCh := make(chan error)
 
        for _, fd := range listenerFds {
-               go concurrentlyRegister(e, fd, errCh)
+               // concurrentlyRegister
+               go func(e *epoll.Epoll, fd int, errCh chan error) {
+                       err := e.Register(fd, epoll.Readable)
+                       errCh <- err
+               }(e, fd, errCh)
        }
 
        for range listenerFds {
```
but maybe that is actually a bit ugly :/ and the name is already clear enough - nevermind I guess :)

Reply:
I think I'll leave it as is... my gut (python) reaction is that calling an inline-defined function is slower than calling a normal function since the function has to be dynamically parsed and allocated, but this probably isn't a problem with a compiled language. But I know it will be fast as-is :)